gear/csv_to_json_transformer/src/python/csv_app/run.py

```diff
@@ -0,0 +1,95 @@
+from flywheel_gear_toolkit import GearToolkitContext
+from gear_execution.gear_execution import (
+    ClientWrapper,
+    ContextClient,
+    GearEngine,
+    GearExecutionEnvironment,
+    GearExecutionError,
+    InputFileWrapper,
+)
+from inputs.parameter_store import ParameterStore
+from outputs.errors import ListErrorWriter
+
+from csv_app.main import run
+
+logging.basicConfig(stream=sys.stdout, level=logging.INFO)
+log = logging.getLogger(__name__)
+
+
+class CsvToJsonVisitor(GearExecutionEnvironment):
+    """The gear execution visitor for the csv-to-json-transformer app."""
+
+    def __init__(self, client: ClientWrapper,
+                 file_input: InputFileWrapper) -> None:
+        self.__client = client
+        self.__file_input = file_input
+
+    @classmethod
+    def create(
+            cls, context: GearToolkitContext,
+            parameter_store: Optional[ParameterStore]) -> 'CsvToJsonVisitor':
+        client = ContextClient.create(context=context)
+        file_input = InputFileWrapper.create(input_name='input_file',
+                                             context=context)
+
+        return CsvToJsonVisitor(client=client, file_input=file_input)
+
+    def run(self, context: GearToolkitContext) -> None:
+        """Runs the CSV to JSON Transformer app.
+
+        Args:
+          context: the gear execution context
+        """
+
+        proxy = self.__client.get_proxy()
+        file_id = self.__file_input.file_id
+        try:
+            file = proxy.get_file(file_id)
+        except ApiException as error:
+            raise GearExecutionError(
+                f'Failed to find the input file: {error}') from error
+
+        project = proxy.get_project_by_id(file.parents.project)
+        if not project:
+            raise GearExecutionError(
+                f'Failed to find the project with ID {file.parents.project}')
+
+        with open(self.__file_input.filepath, mode='r',
+                  encoding='utf-8') as csv_file:
+            error_writer = ListErrorWriter(container_id=file_id,
+                                           fw_path=proxy.get_lookup_path(file))
+            success, sys_errors = run(input_file=csv_file,
+                                      proxy=proxy,
+                                      project=project,
+                                      error_writer=error_writer)
+
+            context.metadata.add_qc_result(self.__file_input.file_input,
+                                           name='validation',
+                                           state='PASS' if success else 'FAIL',
+                                           data=error_writer.errors())
+
+            context.metadata.add_file_tags(self.__file_input.file_input,
+                                           tags=context.manifest.get(
+                                               'name',
+                                               'csv-to-json-transformer'))
+            if sys_errors:
```

Comment:
it seems weird to me to have the gear add qc results and file tags and *then* exit with an error. If it is OK to keep executing after the error occurs, just log it. Otherwise, raise an exception and make sure it gets caught and transformed into a GearExecutionError.

Reply:
Is it ok to raise an exception and stop processing the reset of the visits if a system error occurred while processing a visit?
If we just log and proceed, these errors will go unnoticed unless someone explicitly checks the log.

Reply:
What is the chance that one file has a system error on save and the rest don't also fail?

An alternative would be to log it and send an email to nacc_dev@uw.edu